### PR TITLE
Users/stevenh/zosmf utils

### LINKED ISSUE #1124: zowe-configure-instance.sh - Errors in sed
**Describe the bug**
instance.env file does not get created 

**To Reproduce**
1. Install Zowe 1.8 SMPE Version
2. Run bin/zowe-configure-instance.sh -c /var/zowe/1.8.0
3. See error : sed: FSUM7296 newline or end of file found in pattern

**Expected behavior**
File instance.env gets populated.

**Screenshots**
/var/usr/lpp/zowe >bin/zowe-configure-instance.sh -c /var/zowe/1.8.0
Creating zowe instance in /var/zowe/1.8.0
Checking existing /var/zowe/1.8.0/instance.env for updated properties
sed: FSUM7296 newline or end of file found in pattern
Missing properties that will be appended to /var/zowe/1.8.0/instance.env:

Properties added, please review these before starting zowe.
Configure instance completed. Please now review the properties in /var/zowe/1.8.0/instance.env to check they are correct.
To start Zowe run the script /var/zowe/1.8.0/bin/zowe-start.sh
   (or in SDSF directly issue the command /S ZWESVSTC,INSTANCE='/var/zowe/1.8.0')
To stop Zowe run the script /var/zowe/1.8.0/bin/zowe-stop.sh
  (or in SDSF directly the command /C ZWESVSTC)

**Desktop (please complete the following information):**
 - OS: Linux
 - Browser Chrome
 - Version 2.3

**Additional context**
Running under z/OS 2.4

### PR DESCRIPTION
**Pull request by stevenhorsman**
#### PR type
- [X] Bugfix

#### Relevant issues
Depends on #1400 
fixes #1124 

#### Changes proposed in this PR
Add new code that tightens zosmf port detection to deal better with conflict with the ZERT port

### CODE REVIEW

**Comment by stevenhorsman** [comment]
Convenience and smpe tests:
https://wash.zowe.org:8443/view/Zowe/job/zowe-install-test/view/change-requests/job/PR-1430/2/
https://wash.zowe.org:8443/view/Zowe/job/zowe-install-test/view/change-requests/job/PR-1430/3/

File: bin/utils/zosmf-utils.sh [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    return 1
+  fi
+  export ZOWE_ZOSMF_PORT=${zosmf_port_list}
+}
+
+# Expects the following to be exported
+# - ZOSMF_PORT - The SSL port z/OSMF is listening on.
+# - ZOSMF_HOST - The hostname, or IP Address z/OSMF can be reached
+validate_zosmf_host_and_port() {
+  if [[ -z "${ZOSMF_HOST}" || -z "${ZOSMF_PORT}" ]]
```
Comment (jackjia-ibm) [inline bin/utils/zosmf-utils.sh:86]: I think pass from parameter looks better like this `validate_zosmf_host_and_port <host> <port>`. This makes this function more independent, less side effects.

File: bin/utils/common.sh [hunk 2]
```diff
@@ -10,7 +10,7 @@
 # Copyright IBM Corporation 2020
 ################################################################################
 
-#TODO LATER - do we want to provide a ENV_VAR flag that toggles whether errors are printed or not?
+ZWE_PRINT_ERRORS="true" # Override in scripts to prevent errors being printed - don't forget to reset after!
```
[the remark this answers is not in the document]
Comment (stevenhorsman) [reply bin/utils/common.sh]: Removed this code - I was trying to provide a way for multiple checks to be run without error-ing, but I think you are correct and that just making them route to /dev/null will be simpler

File: bin/utils/zosmf-utils.sh [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  else
+    number_of_matches=$(echo "${zosmf_port_list}" | wc -l)
+  fi
+  
+  # Zip 1124 - in z/OS 2.4 some people have reported 2 ports for z/OS MF. If we know host name at this point, try to work out which is correct
+  if [[ ${number_of_matches} -gt 1 ]] && [[ -n "${ZOSMF_HOST}" ]] && [[ -n "${NODE_HOME}" ]];
+  then
+    OLDIFS="$IFS"
+    IFS="
+    "
```
[the remark this answers is not in the document]
Comment (stevenhorsman) [reply bin/utils/zosmf-utils.sh:62]: So I'm not sure I fully understand this. The defect that the customer found was that we could get multiple ports for zosmf. So `zosmf_port_list=$(onetstat -b -E IZUSVR1 2>/dev/null | grep .*Listen | awk '{ print $4 }')` gives ```
20015
01443
```
I tried multiple ways of splitting a variables based on newline and the IFS method was the only one I got to work. Do you know of a better way to do this then?

File: bin/utils/zosmfHttpRequest.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+const https = require('https');
```
[the remark this answers is not in the document]
Comment (stevenhorsman) [reply bin/utils/zosmfHttpRequest.js]: Yep - fully migrated to the new version now

File: bin/utils/zosmf-utils.sh [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  then 
+    print_error_message "ZOSMF_HOST and ZOSMF_PORT are not both set"
+    return 1
+  else
+    # SH: Note - if node is not available then will continue with a warning
+    if [ -z "${NODE_HOME}" ];
+    then
+      log_message "Warning: Could not validate if z/OS MF is available on 'https://${ZOSMF_HOST}:${ZOSMF_PORT}/zosmf/info'"
+    else
+      http_response_code=$(${NODE_HOME}/bin/node ${ROOT_DIR}/scripts/utils/zosmfHttpRequest.js ${ZOSMF_HOST} ${ZOSMF_PORT})
```
[the remark this answers is not in the document]
Comment (stevenhorsman) [reply bin/utils/zosmf-utils.sh]: Yep - fully migrated to the new version now

File: bin/utils/zosmf-utils.sh (hunk 1 above)
Comment (stevenhorsman) [reply bin/utils/zosmf-utils.sh]: Updated

File: bin/utils/zosmf-utils.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (jackjia-ibm) [reply bin/utils/zosmf-utils.sh:62]: IFS default already has line feed in it, also has tab and space. So if we just want to split with line feed, default values should be good enough. (Except some other scripts changed IFS value but didn't revert it.)

I tried this:

```
zosmf_port_list=$(onetstat -b -E ZWE1EF | grep ClosWait | awk '{print $4 }')
for port in ${zosmf_port_list}; do
  echo "[$port]"
done
```

and I got port one by one. Please let me know if I test this correctly or not, or if I misunderstand the issue.

[the remark this answers is not in the document]
Comment (stevenhorsman) [reply bin/utils/zosmf-utils.sh]: Nope - I think of the several ways I tried this wasn't one! Seems to work fine for me.

**Comment by stevenhorsman** [comment]
Updated tests:
https://wash.zowe.org:8443/view/Zowe/job/zowe-install-test/view/change-requests/job/PR-1430/5/
https://wash.zowe.org:8443/view/Zowe/job/zowe-install-test/view/change-requests/job/PR-1430/6/

File: bin/utils/zosmf-utils.sh [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    for port in ${zosmf_port_list}; do
+      http_response_code=$(${NODE_HOME}/bin/node ${utils_dir}/zosmfHttpRequest.js ${ZOSMF_HOST} ${port})
+      if [[ ${http_response_code} == 200 ]]
+      then
+        zosmf_port_list=${port}
+        number_of_matches=1
+        break
+      fi
+    done
+    export IFS="$OLDIFS"
```
Comment (jackjia-ibm) [inline bin/utils/zosmf-utils.sh:75]: Will need to remove this line since `$OLDIFS` is empty.

File: scripts/utils/validate-zosmf-host-and-port.sh [hunk 7]
```diff
@@ -7,30 +7,10 @@
...
-        . ${ROOT_DIR}/scripts/utils/error.sh "Could not contact z/OS MF on 'https://${ZOSMF_HOST}:${ZOSMF_PORT}/zosmf/info' - $RESPONSE_CODE"
-      fi
-    fi
-  else
-    echo "Warning: Could not validate if z/OS MF is available on 'https://${ZOSMF_HOST}:${ZOSMF_PORT}/zosmf/info'"
-  fi
-fi
\ No newline at end of file
+. ${ROOT_DIR}/bin/utils/zosmf-utils.sh
+"${ZOSMF_HOST}" "${ZOSMF_PORT}"
```
Comment (jackjia-ibm) [inline scripts/utils/validate-zosmf-host-and-port.sh:16]: This probably misses `validate_zosmf_host_and_port` function name.

**Comment by stevenhorsman** [comment]
> Sorry 2 minor issues.

No need to apologise - sorry I made such a poor job of it. Thanks for the excellently thorough review!

**Comment by stevenhorsman** [comment]
New tests: https://wash.zowe.org:8443/view/Zowe/job/zowe-install-test/view/change-requests/job/PR-1430/7/ https://wash.zowe.org:8443/view/Zowe/job/zowe-install-test/view/change-requests/job/PR-1430/8/

**Pull request merged**